src/cocotb_tools/pytest/_init.py

```diff
@@ -0,0 +1,99 @@
+        *env.as_args("COCOTB_PYTEST_ARGS"),
+        # Node identifier of cocotb runner
+        nodeid=env.as_str("COCOTB_PYTEST_NODEID"),
+        # List of cocotb runner keywords
+        keywords=env.as_list("COCOTB_PYTEST_KEYWORDS"),
+        # Provide list of test modules (Python modules with cocotb tests) to be loaded
+        test_modules=env.as_list("COCOTB_TEST_MODULES"),
+        # Cocotb runner is using generated JUnit XML results file to determine
+        # if executed cocotb tests passed or failed. Test function (cocotb runner)
+        # from the main pytest parent process will also fail if any of cocotb test failed.
+        xmlpath=env.as_str("COCOTB_RESULTS_FILE"),
+        # Path to directory location from where pytest was invoked
+        invocation_dir=env.as_path("COCOTB_PYTEST_DIR"),
+        # IPC address (Unix socket, Windows pipe, TCP, ...) to tests reporter
+        reporter_address=env.as_str("COCOTB_PYTEST_REPORTER_ADDRESS"),
+        # Name of HDL top level design
+        toplevel=env.as_str("COCOTB_TOPLEVEL"),
+    )
+
+    cocotb._regression_manager = cast("cocotb.regression.RegressionManager", manager)
+    cocotb._regression_manager.start_regression()
+
+
+def _setup_simulation_environment(argv: list[str] | None = None) -> None:
+    """Setup minimal required simulation environment for pytest and cocotb."""
+    cocotb.log = getLogger("test")
+    cocotb.simulator.set_sim_event_callback(_sim_event)
+    _setup_gpi_logger()
+
+    cocotb.argv = argv or []
+    cocotb.is_simulation = True
+
+    # sys.path normally includes "" (the current directory), but does not appear to when python is embedded.
+    # Add it back because users expect to be able to import files in their test directory.
+    sys.path.insert(0, "")
+
+    cocotb.RANDOM_SEED = env.as_int("COCOTB_RANDOM_SEED", int(time()))
+    cocotb.SIM_NAME = simulator.get_simulator_product().strip()
+    cocotb.SIM_VERSION = simulator.get_simulator_version().strip()
+
+    _process_plusargs()
+    _process_packages()
+    _setup_root_handle()
+
+    simtime._init()
+    seed(cocotb.RANDOM_SEED)
+
+
+def _setup_root_handle() -> None:
+    """Function that will set the :py:var:`cocotb.top` handler."""
+    root_name: str = env.as_str("COCOTB_TOPLEVEL")
+
+    if "." in root_name:
+        # Skip any library component of the toplevel
+        root_name = root_name.split(".", 1)[1]
+
+    handle = simulator.get_root_handle(root_name)
+
+    if handle:
+        cocotb.top = cocotb.handle._make_sim_object(handle)
+
+    # If None, pytest will raise an exception
+
+
+def _sim_event(msg: str) -> None:
+    """Function that can be called externally to signal an event."""
+    if hasattr(cocotb, "_regression_manager"):
+        cocotb._regression_manager._fail_simulation(msg)
+    else:
+        # Early stage when logging is not yet configured by pytest logging plugin
+        print(msg, file=sys.stderr, flush=True)
```

Comment:
Note to self, all of this could be eliminated by refactoring init behavior.

Reply:
@ktbarrett I have just synced with cocotb `master` branch. And removed `msg` argument and `print` statement. I also renamed `_fail_simulation(msg)` method to `_on_sim_end()` method in this commit 8e43c242f930597adcc097b1db105f31e5b2a342